src/API/QueryFieldService.php

```diff
@@ -0,0 +1,103 @@
+ */
+namespace EzSystems\EzPlatformQueryFieldType\API;
+
+use eZ\Publish\API\Repository\ContentTypeService;
+use eZ\Publish\API\Repository\SearchService;
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\Content\ContentInfo;
+use eZ\Publish\API\Repository\Values\Content\Search\SearchHit;
+use eZ\Publish\API\Repository\Values\ContentType\FieldDefinition;
+use eZ\Publish\Core\QueryType\QueryTypeRegistry;
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * Executes a query and returns the results.
+ */
+class QueryFieldService
+{
+    /** @var \eZ\Publish\Core\QueryType\QueryTypeRegistry */
+    private $queryTypeRegistry;
+
+    /** @var \eZ\Publish\API\Repository\SearchService */
+    private $searchService;
+
+    /** @var \eZ\Publish\API\Repository\ContentTypeService */
+    private $contentTypeService;
+
+    public function __construct(
+        SearchService $searchService,
+        ContentTypeService $contentTypeService,
+        QueryTypeRegistry $queryTypeRegistry
+    ) {
+        $this->searchService = $searchService;
+        $this->contentTypeService = $contentTypeService;
+        $this->queryTypeRegistry = $queryTypeRegistry;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\Content\Content $content
+     * @param string $fieldDefinitionIdentifier
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Content[]
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function loadFieldData(Content $content, string $fieldDefinitionIdentifier): array
+    {
+        $fieldDefinition = $this->getFieldDefinition($content->contentInfo, $fieldDefinitionIdentifier);
+        $queryType = $this->queryTypeRegistry->getQueryType($fieldDefinition->fieldSettings['QueryType']);
+        $parameters = $this->resolveParameters($fieldDefinition->fieldSettings['Parameters'], $content);
+
+        return array_map(
+            function (SearchHit $searchHit) {
+                return $searchHit->valueObject;
+            },
+            $this->searchService->findContent($queryType->getQuery($parameters))->searchHits
+        );
+    }
+
+    public function getFieldDefinition(ContentInfo $contentInfo, string $fieldDefinitionIdentifier): FieldDefinition
+    {
+        return $queryFieldDefinition =
+            $this
+                ->contentTypeService->loadContentType($contentInfo->contentTypeId)
+                ->getFieldDefinition($fieldDefinitionIdentifier);
+    }
+
+    /**
+     * @param string $parameters parameters in JSON format
+     * @param \eZ\Publish\API\Repository\Values\Content\Content $content
+     *
+     * @return array
+     */
+    private function resolveParameters(string $parametersString, Content $content): array
+    {
+        $parameters = [];
+        foreach (Yaml::parse($parametersString) as $key => $parameter) {
```

Comment:
I'm not sure if YAML is a good idea. Q: Also what about error handling?

Reply:
What would you suggest ? It used to be json, but YAML *is* _better than JSON for this context_. 

The YAML string is validated in `EzSystems\EzPlatformQueryFieldType\FieldType\Query\Type::validateFieldSettings()`.

Reply:
I have changed how that config is made:
- input is Yaml, with validation on submit
- storage is json, handled by the legacy converter